src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -386,6 +386,64 @@ public void TrimExcess()
             }
         }
 
+        /// <summary>
+        /// Ensures that the capacity of this Queue is at least the specified <paramref name="capacity"/>.
+        /// </summary>
+        /// <param name="capacity">The minimum capacity to ensure</param>
+        public int EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), capacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            if (_array.Length < capacity)
+            {
+                int newCapacity = (int)((long)(_array.Length == 0 ? 1 : _array.Length) * (long)GrowFactor / 100);
+                while (newCapacity < capacity)
+                {
+                    newCapacity = (int)((long)newCapacity * (long)GrowFactor / 100);
+                }
+
+                // MaxArrayLength is defined in Array.MaxArrayLength and in gchelpers in CoreCLR.
+                // It represents the maximum number of elements that can be in an array where
+                // the size of the element is greater than one byte; a separate, slightly larger constant,
+                // is used when the size of the element is one.
+                const int MaxArrayLength = 0x7FEFFFFF;
+                if ((uint)newCapacity > MaxArrayLength)
+                {
+                    newCapacity = MaxArrayLength;
+                    if (newCapacity < capacity) newCapacity = capacity;
+                }
+
+                if (newCapacity < _array.Length + MinimumGrow)
+                {
+                    newCapacity = _array.Length + MinimumGrow;
+                }
+
+                T[] newArray = new T[newCapacity];
+                if (_size > 0)
+                {
+                    if (_head < _tail)
+                    {
+                        Array.Copy(_array, _head, newArray, 0, _size);
+                    }
+                    else
+                    {
+                        Array.Copy(_array, _head, newArray, 0, _array.Length - _head);
+                        Array.Copy(_array, 0, newArray, _array.Length - _head, _tail);
+                    }
+                }
+
+                _array = newArray;
+                _head = 0;
+                _tail = _size;
+                _version++;
```

Comment:
Should resize operations invalidate existing enumerators? Unless the collection is modified concurrently I don't see why it could cause issues cc @stephentoub

Reply:
Checked code, maybe because EnsureCapacity() renew internal buffer instance so existing code treats it as version (data) update. MoveNext checked version (data) changed so it blocked current Enumerator.MoveNext ? Seems Dictionary.[EnsureCapacity ](https://github.com/dotnet/runtime/blob/bee5b795add1a3113925be13d25a0d419ff0d9c4/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1037)and Dictionary.[Enumerator ](https://github.com/dotnet/runtime/blob/bee5b795add1a3113925be13d25a0d419ff0d9c4/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L1274)has same case and its test also verified it can [invalidate ](https://github.com/dotnet/runtime/blob/bee5b795add1a3113925be13d25a0d419ff0d9c4/src/libraries/System.Collections/tests/Generic/Dictionary/Dictionary.Generic.Tests.cs#L285)enumeration. So we should turn to check they do invalidate enumeration ?